power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp

```diff
@@ -714,52 +766,174 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
+                    auto const* id_buffer = comp_buffer.template get_col_data_at_index<ID>(id_idx);
+
+                    result.ids_all_na =
+                        std::all_of(id_buffer, id_buffer + comp_info.total_elements, [](ID id) { return is_nan(id); });
+                    result.ids_part_na = std::any_of(id_buffer, id_buffer + comp_info.total_elements,
+                                                     [](ID id) { return is_nan(id); }) &&
+                                         !result.ids_all_na;
+                    assert(result.elements_in_base > 0);
+                    auto const n_scenario_comp_in_update =
+                        result.elements_in_base * n_scenarios > comp_info.total_elements
+                            ? comp_info.total_elements / result.elements_in_base
+                            : n_scenarios;
+                    result.ids_match = [id_buffer, elements_ps = result.elements_in_base, n_scenario_comp_in_update]() {
+                        bool all_match = true;
+                        for (Idx i = 0; i < elements_ps; ++i) {
+                            for (Idx j = 0; j < n_scenario_comp_in_update; ++j) {
+                                if (id_buffer[i] != id_buffer[i + elements_ps * j]) {
+                                    all_match = false;
+                                    break;
+                                }
+                            }
+                            if (!all_match) {
+                                break;
+                            }
+                        }
+                        return all_match;
+                    }();
+                    if (comp_info.elements_per_scenario != invalid_index) {
+                        result.elements_ps_in_update = comp_info.elements_per_scenario;
+                    }
+                } else {                        // no id,
+                    result.ids_all_na = true;   // no id, all NA
+                    result.ids_part_na = false; // no id, no part NA
+                    result.ids_match = true;    // no id, all match
+                }
+            } else {
+                process_buffer_span.template operator()<CT>(
+                    update_data.get_buffer_span_all_scenarios<meta_data::update_getter_s, CT>(), result);
+            }
+            if (comp_index >= 0 && result.uniform) {
+                result.elements_ps_in_update =
+                    update_data.get_component_info(comp_index).elements_per_scenario; // -1 for sparse
+            }
+            return result;
         };
 
-        // check all components
-        auto const update_independent = run_functor_with_all_types_return_array(is_component_update_independent);
-        return std::ranges::all_of(update_independent, [](bool const is_independent) { return is_independent; });
+        // check and return indenpendence of all components
+        return run_functor_with_all_types_return_vector(check_each_component);
```

Comment:
i think `array` would've still worked because the amount of component types is determined compile-time, right? or is this done on a runtime basis?

Reply:
Managing the compile time array size and carry it across many functions is not so easily achievable with an `auto`.

Reply:
hmm... let's try to address this in the refactoring then.

Reply:
Agreed to do it in the refactoring.